type [1073741824]uint16 too large

`set GOARCH=386`
`\example_glfw_opengl3> go build -v -tags 'glfw' -o 32.exe .`


D:\Golib\pkg\mod\github.com\inkyblackness\imgui-go\v2@v2.1.0\AllocatedGlyphRanges.go:32:16: type [1073741824]uint16 too large
D:\Golib\pkg\mod\github.com\inkyblackness\imgui-go\v2@v2.1.0\GlyphRanges.go:24:16: type [1073741824]uint16 too large

Assuming your goal is to get a binary for an Intel 386 CPU and you are wondering why it does not work: I find your enthusiasm great, and I also wonder how you got that far.

Because when I try to run this, go bails out earlier, as the libraries won't support this:
```
GOARCH=386 go run -tags 'glfw' .
../../internal/renderers/OpenGL2.go:7:2: build constraints exclude all Go files in /home/dertseha/go/pkg/mod/github.com/go-gl/gl@v0.0.0-20190320180904-bf2b1f2f34d7/v2.1/gl
../../internal/renderers/OpenGL3.go:7:2: build constraints exclude all Go files in /home/dertseha/go/pkg/mod/github.com/go-gl/gl@v0.0.0-20190320180904-bf2b1f2f34d7/v3.2-core/gl
../../internal/platforms/glfw.go:10:2: build constraints exclude all Go files in /home/dertseha/go/pkg/mod/github.com/go-gl/glfw/v3.3/glfw@v0.0.0-20200222043503-6f7a984d4dc4
```

This is running on Go 1.14 (a similar error was happening with 1.13, albeit a little more cryptic) - both for Linux and MS Windows and trying it with newest libraries.

What kind of  setup are you running this in? If you can reproduce this, I'd also be happy with a pull-request.
It comes down to a common hack to get back from a raw C-pointer to a typed-array of a certain size.  The 386 build would need a different arbitrary large-enough size for the base array pointer.

--- a/AllocatedGlyphRanges.go
+++ b/AllocatedGlyphRanges.go
@@ -31,7 +31,7 @@ func (builder *GlyphRangesBuilder) Build() AllocatedGlyphRanges {
 	const uint16PerRangeEntry = 2
 	ranges := builder.mergedRanges()
 	raw := C.malloc(C.size_t(bytesPerUint16 * ((len(ranges) * uint16PerRangeEntry) + 1)))
-	rawSlice := (*[1 << 30]uint16)(raw)[:]
+	rawSlice := ptrToUint16Slice(raw)
 	outIndex := 0
 	for _, r := range ranges {
 		rawSlice[outIndex+0] = r.from

--- a/FontAtlas.go
+++ b/FontAtlas.go
@@ -103,7 +103,7 @@ func (atlas FontAtlas) AddFontFromMemoryTTFV(
 	}
 
 	fontDataC := C.malloc(C.size_t(len(fontData)))
-	cBuf := (*[1 << 30]byte)(fontDataC)
+	cBuf := ptrToByteSlice(fontDataC)
 
 	copy(cBuf[:], fontData)
 

--- a/FontAtlas.go
+++ b/FontAtlas.go
@@ -105,7 +105,7 @@ func (atlas FontAtlas) AddFontFromMemoryTTFV(
 	fontDataC := C.malloc(C.size_t(len(fontData)))
 	cBuf := ptrToByteSlice(fontDataC)
 
-	copy(cBuf[:], fontData)
+	copy(cBuf, fontData)
 
 	fontHandle := C.iggAddFontFromMemoryTTF(atlas.handle(), (*C.char)(fontDataC), C.int(len(fontData)), C.float(sizePixels),
 		config.handle(), glyphRange.handle())

--- a/GlyphRanges.go
+++ b/GlyphRanges.go
@@ -21,7 +21,7 @@ func (glyphs GlyphRanges) extract() (result []glyphRange) {
 	if glyphs == 0 {
 		return
 	}
-	rawSlice := (*[1 << 30]uint16)(unsafe.Pointer(glyphs.handle()))[:]
+	rawSlice := ptrToUint16Slice(unsafe.Pointer(glyphs.handle()))
 	index := 0
 	// iterate until end of list or an arbitrary paranoia limit, should the list not be proper.
 	for (rawSlice[index] != 0) && (index < 1000) {

--- a/InputTextCallbackData.go
+++ b/InputTextCallbackData.go
@@ -125,7 +125,7 @@ func (data InputTextCallbackData) Buffer() []byte {
 		return nil
 	}
 	textLen := data.bufTextLen()
-	return ((*[1 << 30]byte)(unsafe.Pointer(ptr)))[:textLen]
+	return ptrToByteSlice(unsafe.Pointer(ptr))[:textLen]
 }
 
 // MarkBufferModified indicates that the content of the buffer was modified during a callback.

--- a/WrapperConverter.go
+++ b/WrapperConverter.go
@@ -77,8 +77,9 @@ func newStringBuffer(initialValue string) *stringBuffer {
 	bufSize := len(rawText) + 1
 	newPtr := C.malloc(C.size_t(bufSize))
 	zeroOffset := bufSize - 1
-	copy(((*[1 << 30]byte)(newPtr))[:zeroOffset], rawText)
-	((*[1 << 30]byte)(newPtr))[zeroOffset] = 0
+	buf := ptrToByteSlice(newPtr)
+	copy(buf[:zeroOffset], rawText)
+	buf[zeroOffset] = 0
 
 	return &stringBuffer{ptr: newPtr, size: bufSize}
 }
@@ -101,7 +102,7 @@ func (buf *stringBuffer) resizeTo(requestedSize int) {
 	if copySize > 0 {
 		C.memcpy(newPtr, buf.ptr, C.size_t(copySize))
 	}
-	((*[1 << 30]byte)(newPtr))[bufSize-1] = 0
+	ptrToByteSlice(newPtr)[bufSize-1] = 0
 	C.free(buf.ptr)
 	buf.ptr = newPtr
 	buf.size = bufSize
@@ -111,6 +112,21 @@ func (buf stringBuffer) toGo() string {
 	if (buf.ptr == nil) || (buf.size < 1) {
 		return ""
 	}
-	((*[1 << 30]byte)(buf.ptr))[buf.size-1] = 0
+	ptrToByteSlice(buf.ptr)[buf.size-1] = 0
 	return C.GoString((*C.char)(buf.ptr))
+}
+
+// unrealisticLargePointer is used to cast an arbitrary native pointer to a slice.
+// Its value is chosen to fit into a 32bit architecture, and still be large
+// enough to cover "any" data blob. Note that this value is in bytes.
+// Should an array of larger primitives be addressed, be sure to divide the value
+// by the size of the elements.
+const unrealisticLargePointer = 1 << 30
+
+func ptrToByteSlice(p unsafe.Pointer) []byte {
+	return (*[unrealisticLargePointer]byte)(p)[:]
+}
+
+func ptrToUint16Slice(p unsafe.Pointer) []uint16 {
+	return (*[unrealisticLargePointer / 2]uint16)(p)[:]
 }

